writer/writer_test.go

```diff
@@ -25,33 +30,210 @@ import (
+	defer os.RemoveAll(updateTestDir)
+	err := MakeFakeUpdateDir(updateTestDir, dirStructInvalid)
+	assert.NoError(t, err)
+
+	artifactWriter := ArtifactsWriter{
+		updateLocation:  updateTestDir,
+		headerStructure: metadata.MetadataArtifactHeader{Artifacts: ArtifactsHeaderFormat},
+	}
+	err = artifactWriter.Write()
+	assert.Error(t, err)
+}
+
+func TestGenerateHash(t *testing.T) {
+	tempDir, _ := ioutil.TempDir("", "update")
+	defer os.RemoveAll(tempDir)
+
+	err := MakeFakeUpdateDir(tempDir,
+		[]metadata.MetadataDirEntry{
+			{Path: "update.ext4", IsDir: false},
+			{Path: "next_update.ext3", IsDir: false},
+		})
+	assert.NoError(t, err)
+
+	err = ioutil.WriteFile(filepath.Join(tempDir, "update.ext4"),
+		[]byte("file content"), os.ModePerm)
+	assert.NoError(t, err)
+	err = ioutil.WriteFile(filepath.Join(tempDir, "next_update.ext3"),
+		[]byte("different file content"), os.ModePerm)
+	assert.NoError(t, err)
+
+	updates, err := ioutil.ReadDir(tempDir)
+	assert.NoError(t, err)
+	assert.NotNil(t, updates)
+
+	artifactWriter := ArtifactsWriter{}
+
+	upd := updateArtifact{path: filepath.Join(tempDir, "update.ext4")}
+	err = artifactWriter.calculateChecksum(&upd)
+	assert.NoError(t, err)
+	assert.Equal(t, []byte("e0ac3601005dfa1864f5392aabaf7d898b1b5bab854f1acb4491bcd806b76b0c"), upd.checksum)
+
+	upd = updateArtifact{path: filepath.Join(tempDir, "next_update.ext3")}
+	err = artifactWriter.calculateChecksum(&upd)
+	assert.NoError(t, err)
+	assert.Equal(t, []byte("90094b71a0bf15ee00e087a3be28579483fb759a718fa4ca97be215b42021121"), upd.checksum)
+
+	upd = updateArtifact{path: filepath.Join(tempDir, "non_existing")}
+	err = artifactWriter.calculateChecksum(&upd)
+	assert.Error(t, err)
+	assert.Empty(t, upd.checksum)
+}
+
+var dirStructOK = []metadata.MetadataDirEntry{
+	{Path: "0000", IsDir: true},
+	{Path: "0000/data", IsDir: true},
+	{Path: "0000/data/update.ext4", IsDir: false},
+	{Path: "0000/data/update_next.ext3", IsDir: false},
+	{Path: "0000/type-info", IsDir: false},
+	{Path: "0000/meta-data", IsDir: false},
+	{Path: "0000/signatures", IsDir: true},
+	{Path: "0000/signatures/update.sig", IsDir: false},
+	{Path: "0000/signatures/update_next.sig", IsDir: false},
+	{Path: "0000/scripts", IsDir: true},
+	{Path: "0000/scripts/pre", IsDir: true},
+	{Path: "0000/scripts/post", IsDir: true},
+	{Path: "0000/scripts/check", IsDir: true},
+}
+
+var dirStructOKAfterWriting = map[string]metadata.MetadataDirEntry{
+	".":                               {Path: ".", IsDir: true, Required: true},
+	"data":                            {Path: "data", IsDir: true, Required: true},
+	"data/0000.tar.gz":                {Path: "data", IsDir: false, Required: true},
+	"0000/data":                       {Path: "0000/data", IsDir: true, Required: true},
+	"0000/data/update.ext4":           {Path: "0000/data/update.ext4", IsDir: false, Required: true},
+	"0000/data/update_next.ext3":      {Path: "0000/data/update_next.ext3", IsDir: false, Required: true},
```

Comment:
Not really. I think we have agreed on the structure of the directory BEFORE packing it to artifact archive. Data is something that needs to be there. We are not removing it while creating archive. One reason is that we might need to redo the whole thing in case of some errors.